### ISSUE TITLE
custom son serializer to speed up symbol query performance

### ISSUE THREAD

**Issue by martinlippert** [eclipse/lsp4j#299]
While working on our language server I observed that serializing a larger number of symbols to JSON (when sending back the results for the workspace/symbol request) consumes quite a bit of time. I wonder if it would be possible to optimize this serialization specifically for symbols, e.g. by replacing the standard reflective collection serialization with a custom-made serialization for the collection of symbols.

Any idea how this could be implemented?

**Comment by spoenemann** [comment]
Gson allows to annotate a class or a field with `@JsonAdapter` to provide a type adapter factory for instances of that type or field. We already use this at several parts of the protocol to ensure we can parse and serialize correctly according to the spec. I guess this mechanism should work for performance improvement, too.

**Comment by spoenemann** [comment]
I experimented with a custom type adapter. It brought us a little speedup, but not a substantial one.

**Comment by martinlippert** [comment]
Cool to see that you found the time to experiment with that, much appreciated. Is there a way for me to try this out? Is that already available in a CI build?

**Comment by spoenemann** [comment]
It should be available on Sonatype Snapshots.

I also created a Gitpod snapshot for running my little performance measurement:

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io#snapshot/5ae589b3-c706-41ca-a1a3-02f2b802ddc9)

**Comment by martinlippert** [comment]
Yeah, not a huge difference, but a difference... :-) Thanks for taking a look into this and trying this, much much appreciated.
